cdap-ui/app/cdap/components/SecureKeys/index.tsx

```diff
@@ -14,166 +14,181 @@
-import TableCell from '@material-ui/core/TableCell';
-import TableHead from '@material-ui/core/TableHead';
-import TableRow from '@material-ui/core/TableRow';
-import DeleteIcon from '@material-ui/icons/Delete';
-import Button from '@material-ui/core/Button';
-import IconButton from '@material-ui/core/IconButton';
-import Paper from '@material-ui/core/Paper';
-
-interface IState {
-  secureKeys: any[];
-  name: string;
+
+interface ISecureKeyState {
+  createdEpochMs?: number;
   description: string;
+  name: string;
+  properties: Record<string, string>;
   data: string;
 }
 
-// Currently a hidden link to internally manage secure keys
-export default class SecureKeys extends React.PureComponent<{}, IState> {
-  public state = {
-    secureKeys: [],
-    name: '',
-    description: '',
-    data: '',
-  };
+export enum SecureKeyStatus {
+  Normal = 'Normal',
+  Success = 'SUCCESS',
+  Failure = 'FAILURE',
+}
 
-  public componentDidMount() {
-    this.fetchSecureKeys();
+export const initialState = {
+  secureKeys: List([]),
+  secureKeyStatus: SecureKeyStatus.Normal,
+  editMode: false,
+  deleteMode: false,
+  activeKeyIndex: null,
+  loading: true,
+};
+
+export function reducer(state, action) {
+  switch (action.type) {
+    case 'SET_SECURE_KEYS':
+      return { ...state, secureKeys: action.secureKeys };
+    case 'SET_SECURE_KEY_STATUS':
+      return { ...state, secureKeyStatus: action.secureKeyStatus };
+    case 'SET_EDIT_MODE':
+      return { ...state, editMode: action.editMode };
+    case 'SET_DELETE_MODE':
+      return { ...state, deleteMode: action.deleteMode };
+    case 'SET_ACTIVE_KEY_INDEX':
+      return { ...state, activeKeyIndex: action.activeKeyIndex };
+    case 'SET_LOADING':
+      return { ...state, loading: action.loading };
+    default:
+      return state;
   }
+}
 
-  private fetchSecureKeys = () => {
-    const namespace = getCurrentNamespace();
-
-    MySecureKeyApi.list({ namespace }).subscribe((res) => {
-      this.setState({
-        secureKeys: res,
-      });
-    });
-  };
-
-  private handleChange = (property) => {
-    return (e) => {
-      // @ts-ignore
-      this.setState({
-        [property]: e.target.value,
-      });
-    };
+const styles = (): StyleRules => {
+  return {
+    content: {
+      padding: '50px',
+    },
+    loadingBox: {
+      width: '100%',
+      height: '100%',
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'center',
+    },
   };
+};
+
+interface ISecureKeysProps extends WithStyles<typeof styles> {}
+
+const SecureKeysView: React.FC<ISecureKeysProps> = ({ classes }) => {
+  const [state, dispatch] = React.useReducer(reducer, initialState);
+
+  const { secureKeyStatus, editMode, deleteMode, loading } = state;
+
+  const namespace = getCurrentNamespace();
+
+  // Observe `secureKeyStatus` with `useEffect` and forward the value to `secureKeyStatus$`
+  const secureKeyStatusSubject = React.useRef(new BehaviorSubject(secureKeyStatus));
```

Comment:
Can we make this a little simpler?

If I understand correctly,
- We pass `dispatch` of the store to other child components
- On change in any component the component dispatches a status change
- and here we an observable and have a useRef on it for any changes
- and useEffect on the status
- and on each status change we use the observable to refetch the list of secure keys.

Instead of all these can we pass in a callback to all the components (`SecureKeysEdit`, `SecureKeysDelete` and `SecureKeysCreate`), say `onClose` which will be called on unmounting  and here on the function we simply refetch all the keys?

Reply:
This way we don't have to pass in the `dispatch` function. This will be problematic to track since in the future if we want to see where a change is happening. Events can be dispatched by any of the child components for a state maintained in the parent.

Reply:
Resolved! Now dispatch only resides in the main component `SecureKeys/index.tsx`

Reply:
I think you don't need to use observable to watch for changes. The events are deterministic in this case, so you should be able to fetch the the list of secureKeys in a deterministic way on a callback. On update (add, edit, and delete), the modal will call `handleClose`. You can fetch the list of secure keys on the callback being passed to handleClose